src/Features/CSharp/Portable/Completion/CompletionProviders/CompletionUtilities.cs

```diff
@@ -238,4 +240,88 @@ public static TextSpan GetTargetSelectionSpanForInsertedMember(SyntaxNode caretT
                 throw ExceptionUtilities.Unreachable();
         }
     }
+
+    public static bool IsSpeculativeTypeParameterContext(SyntaxTree syntaxTree, int position, SemanticModel? semanticModel, bool inMemberDeclarationOnly, CancellationToken cancellationToken)
+    {
+        var spanStart = position;
+
+        // We could be in the middle of a ref/generic/tuple type, instead of a simple T case.
+        // If we managed to walk out and get a different SpanStart, we treat it as a simple $$T case.
+        while (true)
+        {
+            var oldSpanStart = spanStart;
+
+            spanStart = WalkOutOfGenericType(syntaxTree, spanStart, semanticModel, cancellationToken);
+            spanStart = WalkOutOfTupleType(syntaxTree, spanStart, cancellationToken);
+            spanStart = WalkOutOfRefType(syntaxTree, spanStart, cancellationToken);
```

Comment:
if you make these helpers into local functions, you can just rewrite this as:

spanStart = WalkOutOfGeneric(WalkOutOfTuple(WalkOutOfRef(spanStart)))